### PR TITLE
Add tests for `compare` and `show` commands

### PR DESCRIPTION
**Pull request by mdboom**
The changes to the code itself are only to make things more testable, otherwise, this is just adding tests (not fixing bugs).

Also a small refactoring of the existing tests to reduce duplication.

### CODE REVIEW

File: PORTAL/tests/test___main__.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+| Geometric mean          | (ref)                                                           | 1.31x slower                                                  |
++-------------------------+-----------------------------------------------------------------+---------------------------------------------------------------+
+
+Benchmark hidden because not significant (1): pickle
+Ignored benchmarks (6) of cpython-3.10.4-9d38120e33-fc_linux-b2cf916db80e-pyperformance.json: genshi_text, genshi_xml, gevent_hub, pylint, sqlalchemy_declarative, sqlalchemy_imperative
+    """
+
+    captured = capsys.readouterr()
+
+    assert captured.out.strip().endswith(expected.strip())
```
Comment (ericsnowcurrently) [inline PORTAL/tests/test___main__.py:44]: It may be worth also checking the 3 header lines of the table.  In particular, the filenames should always be the "base" name, rather than the absolute path or any other relative path.

Comment (ericsnowcurrently) [reply PORTAL/tests/test___main__.py:44]: I do see that this would involve more parsing than checking the end lines, so it might not be worth it in this PR.  As an alternative you could add a TODO comment about checking the filenames in the table header.

File: PORTAL/tests/test___main__.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    with pytest.raises(SystemExit) as exc:
+        __main__._parse_and_main(
+            DEFAULT_ARGS + [
+                "show",
+            ],
+            __file__
+        )
+
+    assert exc.value.code == 1
```
Comment (ericsnowcurrently) [inline PORTAL/tests/test___main__.py:60]: Why `1`?  I'd expect the exit code to be `0`.

File: PORTAL/jobs/__main__.py [hunk 3]
```diff
@@ -916,10 +921,15 @@ def handle_args(args, parser):
...
-        argv: Sequence[str] = sys.argv[1:],
-        prog: str = sys.argv[0]
+        argv: Optional[Sequence[str]] = None,
+        prog: Optional[str] = None
 ) -> Tuple[str, Dict[str, Any], str, str, int, str, bool]:
 
+    if argv is None:
+        argv = sys.argv[1:]
+    if prog is None:
+        prog = sys.argv[0]
```
Comment (ericsnowcurrently) [inline PORTAL/jobs/__main__.py:931]: Why is this necessary?  (I'm not opposed, just wondering.)

File: PORTAL/tests/helpers.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from pathlib import Path
+
+
+DATA_ROOT = Path(__file__).parent / "data"
+
+NEW_FILENAME = DATA_ROOT / "cpython-3.11.0b1-8d32a5c8c4-fc_linux-b2cf916db80e.json"
+
+CONFIG_FILENAME = DATA_ROOT / "jobs.json"
```
Comment (ericsnowcurrently) [inline PORTAL/tests/helpers.py:8]: Note that "jobs.json" already has meaning in this project, so consider something like "config.json" instead.

File: PORTAL/tests/data/jobs.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+    "local_user": "benchmarking",
```
Comment (ericsnowcurrently) [inline PORTAL/tests/data/jobs.json:2]: We'd probably need to adjust "local_user" to whoever is running the tests, meaning we'd likely need to generate the config file (at least the first time).

Comment (ericsnowcurrently) [reply PORTAL/tests/data/jobs.json:2]: Also, the config doesn't have "data_dir" set, so commands that use it will fall back to the default (our actual production data).  Instead, consider using a data dir rooted under tests/data.  It may need to be different depending on the test.

Comment (ericsnowcurrently) [reply PORTAL/tests/data/jobs.json:2]: Both of those could probably wait, but it would be worth opening an issue at least.

File: PORTAL/jobs/_pyperformance.py [hunk 6]
```diff
@@ -3245,6 +3248,8 @@ def from_remote(  # type: ignore[override]
     ) -> "PyperfResultsRepo":
         if not remote:
             remote = cls.REMOTE
+        if root is None:
+            root = cls._DEFAULT_ROOT
```
Comment (ericsnowcurrently) [inline PORTAL/jobs/_pyperformance.py:3252]: It may make more sense to look to the config file for the default before falling back to the current behavior (a hard-coded default).  So consider adding a new config option for this instead of the monkey-patching approach.

I guess the default from the config would be passed in rather than the config getting used directly here.

File: PORTAL/jobs/__main__.py [hunk 7]
```diff
@@ -669,6 +669,11 @@ def configure_root_logger(
         *,
         maxlevel: int = logging.CRITICAL,
 ) -> None:
+    # pytest does its own monkey-patching of logging that isn't compatible with
+    # this.
+    if "pytest" in sys.modules:
+        return
+
```
Comment (ericsnowcurrently) [inline PORTAL/jobs/__main__.py:676]: Would it make sense to move this down to right after `#logger.propagate = False`?  Then CLI usage would have things like `-v` respected properly.  That said, we probably also want the formatter applied as well.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
I've left a few comments.

File: PORTAL/tests/test___main__.py (hunk 1 above)
Comment (mdboom) [reply PORTAL/tests/test___main__.py:44]: That's a good callout, and not too hard, so I've added it.

File: PORTAL/tests/test___main__.py (hunk 2 above)
Comment (mdboom) [reply PORTAL/tests/test___main__.py:60]: It's "no job currently running": https://github.com/faster-cpython/tools/blob/main/PORTAL/jobs/__main__.py#L108

It's not a great test -- just a start.  Eventually, we'll also want to test that *if* a job is running, *show* will show it...  But maybe not for this PR.

File: PORTAL/jobs/__main__.py (hunk 3 above)
Comment (mdboom) [reply PORTAL/jobs/__main__.py:931]: This refactor here is to put the `if __name__ == '__main__'` clause in a function so I can run the top level of `__main__.py` providing my own arguments from a test.  I could have put these in the kwarg defaults, but I find that can be the source of subtle bugs if `sys.argv` were mutated and that wouldn't be picked up.

File: PORTAL/tests/helpers.py (hunk 4 above)
Comment (mdboom) [reply PORTAL/tests/helpers.py:8]: I'm not opposed, but I think it's clearer this way, since `jobs.json` is the default name of the config file, and this is a test-specific replacement for it.

File: PORTAL/jobs/_pyperformance.py (hunk 6 above)
Comment (mdboom) [reply PORTAL/jobs/_pyperformance.py:3252]: Yeah, that probably makes sense.  The tests would just set it to another subdirectory in the tree.  Will require some refactoring to pass the cfg here.

File: PORTAL/jobs/__main__.py (hunk 7 above)
Comment (mdboom) [reply PORTAL/jobs/__main__.py:676]: Yeah, that works.  Setting the formatter is a bit trickier, because we'd need to set it on pytest's handler.  Probably doable, but maybe save that until we want to test specific logging messages.

File: PORTAL/tests/data/jobs.json (hunk 5 above)
Comment (mdboom) [reply PORTAL/tests/data/jobs.json:2]: Sorry -- I misspoke -- I did the dynamic `data_dir` thing in PR #108.  Let me move that over here.

File: PORTAL/tests/test___main__.py (hunk 1 above)
Comment (ericsnowcurrently) [reply PORTAL/tests/test___main__.py]: Thanks!

File: PORTAL/jobs/__main__.py (hunk 3 above)
Comment (ericsnowcurrently) [reply PORTAL/jobs/__main__.py:931]: Ah, you're saying `sys.argv` might have been changed between the time the module was loaded and `if __name__ == '__main__':` was executed.  (That's the only difference from the existing code that I see.)  That's a very unlikely race when run as a script.  When we run `main()` in tests we're passing in `argv` anyway, aren't we?  So a race shouldn't matter.  I'm probably still missing something here.

File: PORTAL/tests/helpers.py (hunk 4 above)
Comment (ericsnowcurrently) [reply PORTAL/tests/helpers.py:12]: Oops, I was thinking of something else (queue.json, I think).  I'm on board with matching whatever the file is named normally.

File: PORTAL/jobs/_pyperformance.py (hunk 6 above)
Comment (ericsnowcurrently) [reply PORTAL/jobs/_pyperformance.py]: looks good

File: PORTAL/jobs/__main__.py (hunk 7 above)
Comment (ericsnowcurrently) [reply PORTAL/jobs/__main__.py]: Yeah, let's punt on that.  It mostly only matters if we're checking exactly what a command prints to stdout (via the logger).

File: PORTAL/jobs/__main__.py (hunk 3 above)
Comment (mdboom) [reply PORTAL/jobs/__main__.py:931]: I thinking more that a test calls the top level multiple times -- but on second thought a test is unlikely to mutate `sys.argv`.  So it's probably not so relevant.  I'll remove the handling of `None`.

File: PORTAL/jobs/__main__.py [hunk 8]
```diff
@@ -1173,7 +1178,14 @@ def main(
         run_cmd(jobs, reqid=reqid, **(_cmd_kwargs or {}))
 
 
-if __name__ == '__main__':
-    cmd, cmd_kwargs, cfgfile, user, verbosity, logfile, devmode = parse_args()
+def _parse_and_main(
+    argv: Sequence[str] = sys.argv[-1],
```
Comment (ericsnowcurrently) [inline PORTAL/jobs/__main__.py:1182]:
```suggestion
    argv: Sequence[str] = sys.argv[1:],
```

**Review by ericsnowcurrently — APPROVED** [review]
Aside from one small thing to fix, LGTM.
